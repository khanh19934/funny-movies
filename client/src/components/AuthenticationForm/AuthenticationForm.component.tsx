import * as React from "react";
import {useState} from "react"
import {Button, Form, FormGroup, Input} from "reactstrap";
import {pick} from "ramda"

import * as AuthServices from "../../services/auth.service"
import {saveAccessToken, saveRefreshToken} from "../../services/localStorage.service";
import {actions as authActions, connect} from "../../context/appStore"

interface IProps {
    isAuthenticate: boolean;
    email: string
}

const AuthenticationForm: React.FC<IProps> = (props) => {
    const [email, setEmail] = useState<string>('')
    const [password, setPassword] = useState<string>('')
    const [error, setError] = useState<string>("")

    const onChangeEmail = (e: React.ChangeEvent<HTMLInputElement>) => setEmail(e.target.value)
    const onChangePassword = (e: React.ChangeEvent<HTMLInputElement>) => setPassword(e.target.value)

    const handleSubmitButton = async () => {
        try {
            const payload = {email, password}

            const {token, refreshToken} = await AuthServices.loginUser(payload)

            saveAccessToken(token)
            saveRefreshToken(refreshToken)

            authActions.setAuthenticate(true)
            authActions.setUserEmail(email)
        }catch(e) {
            setError(e.response.data.message)
        }


    }

    return props.isAuthenticate ? <p>Welcome ! {props.email}</p > : (
        <Form inline>
            <FormGroup className="mb-2 mr-sm-2 mb-sm-0">
                <Input aria-label="email-input" type="email" name="email" placeholder="Email" value={email} onChange={onChangeEmail} />
            </FormGroup>
            <FormGroup className="mb-2 mr-sm-2 mb-sm-0">
                <div style={{display: 'flex', flexDirection: 'column'}}>
                    <Input
                        type="password"
                        name="password"
                        placeholder="Password"
                        value={password}
                        onChange={onChangePassword}
                    />
                    {error && error}
                </div>

            </FormGroup>
            <Button onClick={handleSubmitButton}>Login / Register</Button>

        </Form>
    )
}

const enhance = connect(pick(['isAuthenticate', 'email']))

export default enhance(AuthenticationForm)