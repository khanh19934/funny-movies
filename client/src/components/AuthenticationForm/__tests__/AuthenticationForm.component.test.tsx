import React from "react";
import "@testing-library/jest-dom/extend-expect";
import mockAxios from 'jest-mock-axios';

import {loginUser} from "../../../services/auth.service"


import {testSnapshots} from "../../../utils/test.util"

import AuthenticationForm from "../AuthenticationForm.component";
import {render, fireEvent} from "@testing-library/react"

afterEach(() => {
    // cleaning up the mess left behind the previous test
    mockAxios.reset();
});

const consoleError = console.error;
beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation((...args) => {
        if (!args[0].includes('Warning: An update to %s inside a test was not wrapped in act')) {
            consoleError(...args);
        }
    });
});

describe('AuthenticationForm', () => {

    testSnapshots(AuthenticationForm, [
        {
            description: 'should match snapshot', props: {
                isAuthenticate: false,
                email: ''
            }
        },
        {
            description: 'should match snapshot', props: {
                isAuthenticate: true,
                email: 'test@gmail.com'
            }
        }
    ])

    // test('Snapshot', () => {
    //     const {asFragment} = render(<AuthenticationForm/>)
    //
    //     expect(asFragment()).toMatchSnapshot()
    // })


    test('Should have email field', () => {
        const {getByPlaceholderText} = render(<AuthenticationForm/>)
        const emailInput = getByPlaceholderText('Email')
        expect(emailInput).toHaveAttribute('type', 'email')
    })

    test('Should have password field', () => {
        const {getByPlaceholderText} = render(<AuthenticationForm/>)
        const password = getByPlaceholderText('Password')
        expect(password).toHaveAttribute('type', 'password')
    })

    test('Should send form payload to api login', async () => {

        const catchFn = jest.fn()
        const thenFn = jest.fn()

        const {getByPlaceholderText, getByText} = render(<AuthenticationForm/>)
        const emailInput = getByPlaceholderText('Email')
        const password = getByPlaceholderText('Password')
        fireEvent.change(emailInput, {target: {value: 'test@gmail.com'}})
        fireEvent.change(password, {target: {value: 'qwe213'}})
        fireEvent.click(getByText('Login / Register'))

        // @ts-ignore
        loginUser({email: emailInput.value, password: password.value}).then(thenFn).catch(catchFn)

        expect(mockAxios.post).toHaveBeenCalledWith('/login', {
            // @ts-ignore
            email: emailInput.value,
            // @ts-ignore
            password: password.value
        })

        mockAxios.mockResponse({data: {token: ""}})

        expect(catchFn).not.toHaveBeenCalled();


    })

    test('Should so welcome message when user authenticated', () => {
        const email = "test@gmail.com"
        // @ts-ignore
        const {getByText} = render(<AuthenticationForm isAuthenticate={true} email={email}/>)

        expect(getByText(`Welcome ! ${email}`)).toBeInTheDocument()
    })



})

