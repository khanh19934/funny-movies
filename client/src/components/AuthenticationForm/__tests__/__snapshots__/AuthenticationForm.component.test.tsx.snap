// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationForm snapshots should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="form-inline"
  >
    <div
      class="mb-2 mr-sm-2 mb-sm-0 form-group"
    >
      <input
        aria-label="email-input"
        class="form-control"
        name="email"
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <div
      class="mb-2 mr-sm-2 mb-sm-0 form-group"
    >
      <div
        style="display: flex; flex-direction: column;"
      >
        <input
          class="form-control"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="btn btn-secondary"
      type="button"
    >
      Login / Register
    </button>
  </form>
</DocumentFragment>
`;

exports[`AuthenticationForm snapshots should match snapshot 2`] = `<DocumentFragment />`;
